File: tests/shapes.c
#include "shapes.h"

#include <math.h>

float shape_perimeter(Shape shape) {
    switch (shape.kind) {
        case ShapeKindCircle:
            return 2.0f * (float)M_PI * shape.data.circle.radius;
        case ShapeKindRectangle:
            return 2.0f * (shape.data.rectangle.width + shape.data.rectangle.height);
        case ShapeKindTriangle:
            return shape.data.triangle.a + shape.data.triangle.b + shape.data.triangle.c;
        case ShapeKindLine:
            return shape.data.line.length;
    }
    return 0.0f;
}

float shape_area(Shape shape) {
    switch (shape.kind) {
        case ShapeKindCircle:
            return (float)M_PI * shape.data.circle.radius * shape.data.circle.radius;
        case ShapeKindRectangle:
            return shape.data.rectangle.width * shape.data.rectangle.height;
        case ShapeKindTriangle: {
            float s = (shape.data.triangle.a + shape.data.triangle.b + shape.data.triangle.c) / 2.0f;
            return sqrtf(s * (s - shape.data.triangle.a) * (s - shape.data.triangle.b) * (s - shape.data.triangle.c));
        }
        case ShapeKindLine:
            return 0.0f;
    }
    return 0.0f;
}

File: butcher/tests/shapes.nim
type Circle* {.importc, bycopy, header: "tests/shapes.h".} = object
  radius*: cfloat

type Rectangle* {.importc, bycopy, header: "tests/shapes.h".} = object
  width*: cfloat
  height*: cfloat

type Triangle* {.importc, bycopy, header: "tests/shapes.h".} = object
  a*: cfloat
  b*: cfloat
  c*: cfloat

type Line* {.importc, bycopy, header: "tests/shapes.h".} = object
  length*: cfloat

type ShapeKind* {.size: sizeof(cint).} = enum
  ShapeKindCircle,
  ShapeKindRectangle,
  ShapeKindTriangle,
  ShapeKindLine

type ShapeData* {.importc, union, header: "tests/shapes.h".} = object
  circle*: Circle
  rectangle*: Rectangle
  triangle*: Triangle
  line*: Line

type Shape* {.importc, bycopy, header: "tests/shapes.h".} = object
  kind*: ShapeKind
  data*: ShapeData

type ShapeCalculatorFn* = proc(a0: Shape): cfloat {.cdecl.}
type ShapeContainsFn* = proc(a0: Shape, a1: Shape): cint {.cdecl.}
proc shape_perimeter*(shape: Shape): cfloat {.importc, cdecl, header: "tests/shapes.h".}
proc shape_area*(shape: Shape): cfloat {.importc, cdecl, header: "tests/shapes.h".}
proc triangleContainsCircle*(triangle: Triangle, circle: Circle): cint {.importc, cdecl, header: "tests/shapes.h".}
